Documentation for SSO using OpenID connect

apurvabhide opened the pull request:
@rahulbajaj0509 @gnurag can you please provide your inputs for this documentation?

apurvabhide17 committed:
Documentation for SSO using OpenID connect

ekohl commented:
Please also update it in nightly so we don't lose it.

ekohl commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,114 @@
+# export KEYCLOAK_REALM=ssl-realm
+# export KEYCLOAK_USER=admin
+# keycloak-httpd-client-install    \
+--app-name foreman-openidc    \
```
I would indent these to make it clear they're part of the previous command.

ekohl commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,114 @@
+# export KEYCLOAK_URL=https://keycloak.example.com
+# export KEYCLOAK_REALM=ssl-realm
+# export KEYCLOAK_USER=admin
+# keycloak-httpd-client-install    \
```
Does this modify the Apache configuration? If you rerun the installer after, does this revert the installation?

rabajaj0509 replied:
Yes, good catch! once installer is run again it will revert the changes back. I think a puppet module will be required for this, right @ekohl ? Should we add a note here for now or leave that to the handling of puppet module and merge this PR irrespective of that change?

What do you think @ekohl ?

rabajaj0509 commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,105 @@
+### Single Sign-on for Foreman using OpenID Connect protocol
+
+An OpenID Provider(OP)  implements a Single Sign-on (SSO) using  an OpenID Connect (OIDC) protocol. OIDC verifies the identity of the End-user depending on the given authentication details. In this example, [Keycloak](https://www.keycloak.org/) is used as the IDP, and Foreman is the Relying Party (RP)
```
nit: s / `https://keycloak.org` / `https://keycloak.example.com`

ekohl replied:
Actually, I think this is correct since it's referring to the project. There are other OIDC providers.

rabajaj0509 replied:
oh! my bad.

rabajaj0509 commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,105 @@
+
+An OpenID Provider(OP)  implements a Single Sign-on (SSO) using  an OpenID Connect (OIDC) protocol. OIDC verifies the identity of the End-user depending on the given authentication details. In this example, [Keycloak](https://www.keycloak.org/) is used as the IDP, and Foreman is the Relying Party (RP)
+
+IDP provides an ID Token, encoded in JSON Web Token (JWT) to Foreman.
```
Can we use `OpenID provider` here, instead of IDP

rabajaj0509 commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,105 @@
+Prerequisites
+
+* A working installation of Foreman at https://foreman.example.com.
+* A working installation of IDP, for example, Keycloak at https://sso.example.com, which uses OIDC protocol. Ensure that the Keycloak server is running.
```
same. `OpenID provider` instead of IDP

rabajaj0509 commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,105 @@
+
+* A working installation of Foreman at https://foreman.example.com.
+* A working installation of IDP, for example, Keycloak at https://sso.example.com, which uses OIDC protocol. Ensure that the Keycloak server is running.
+* Ensure that both RP and IDP are using https instead of http. If the certificates or the CA is self-signed, ensure that they are added to the End-user certificate trust store.
```
same. `OpenID provider` instead of IDP

rabajaj0509 commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,105 @@
+3. Log in to the Foreman admin page and click **Administer** → **Settings** → **Authentication**.
+ ![Foreman UI](Foreman_UI.png)
+
+4. Set **authorize_login_delegation_auth_source_user_autocreate** to **External** to enable auto-creation of users from external IDP.
```
same. `OpenID provider` instead of IDP

rabajaj0509 reviewed (changes requested):
Just added a few in-line comments but overall this looks great! Thanks for all your efforts @apurvabhide.

apurvabhide commented:
> Please also update it in nightly so we don't lose it.

@ekohl , I am not quite aware of the nightly, could you please elaborate it?

ekohl commented:
> @ekohl , I am not quite aware of the nightly, could you please elaborate it?

You now made all the changes to `/1.24` but we consider `/nightly` the next version. That means for 1.25 we will copy `/nightly` to `/1.25`. If your changes are only in 1.24, it means future releases won't contain it.

apurvabhide commented:
> > @ekohl , I am not quite aware of the nightly, could you please elaborate it?
> 
> You now made all the changes to `/1.24` but we consider `/nightly` the next version. That means for 1.25 we will copy `/nightly` to `/1.25`. If your changes are only in 1.24, it means future releases won't contain it.

@ekohl , I have added it in the `/nightly` and removed it from `/1.24`.

apurvabhide commented:
> > > @ekohl , I am not quite aware of the nightly, could you please elaborate it?
> > 
> > 
> > You now made all the changes to `/1.24` but we consider `/nightly` the next version. That means for 1.25 we will copy `/nightly` to `/1.25`. If your changes are only in 1.24, it means future releases won't contain it.
> 
> @ekohl , I have added it in the `/nightly` and removed it from `/1.24`.

@ekohl, I have added these changes in `/1.24` also. Apologies for the confusion.

ntkathole reviewed (commented):
Overall PR looks good to me! Can we mention about default options to setup in hammer config for oauth/basic auth as well?

ntkathole commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
````diff
@@ -0,0 +1,105 @@
+
+1. Install the mod_auth packages:
+```
+# yum install mod_auth_openidc python2-requests-oauthlib opensc keycloak-httpd-client-install -y
````
opensc is specifically required for smart card authentication. Can we remove this rpm for here and may be include in smartcard auth documentation?

rabajaj0509 replied:
perfect! thanks a lot for reviewing @ntkathole

apurvabhide replied:
> Overall PR looks good to me! Can we mention about default options to setup in hammer config for oauth/basic auth as well?

@ntkathole, Thank you for reviewing, I would like to add hammer configuration options in another PR.

apurvabhide replied:
> opensc is specifically required for smart card authentication. Can we remove this rpm for here and may be include in smartcard auth documentation?

@ntkathole, done, I have removed `opensc` from this command.

ntkathole reviewed (approved):
Thanks @apurvabhide @rahulbajaj0509 Documentation wise ACK. I am leaving syntax and style checking over reviewers.

rabajaj0509 reviewed (approved):
@apurvabhide thanks for all the efforts!

ekohl commented:
This is incompatible with the installer. That doesn't have to block merging documentation but I do think it requires a warning.

apurvabhide commented:
> This is incompatible with the installer. That doesn't have to block merging documentation but I do think it requires a warning.

@ekohl, done, I have added warning for the installer.

ehelms commented:
If I recall, the original plan was for Keycloak users to make use of the Apache puppet modules purge configs being set to false to prevent the installer from clearing this configuration out:

https://github.com/puppetlabs/puppetlabs-apache/blob/31c14cf50251f7692a768b3e3d03d738613baa9b/manifests/init.pp#L294

ekohl commented:
When we want to switch over from passenger to puma we actually rely on that setting to clear up passenger configs. While initially it won't harm them as long as the package is installed, the real cleanup (`yum remove`) will break things. I consider setting that to true an unsupported setup

ekohl commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
````diff
@@ -0,0 +1,106 @@
+
+1. Install the mod_auth packages:
+```
+# yum install mod_auth_openidc python2-requests-oauthlib keycloak-httpd-client-install -y
````
At least on CentOS 7.7 `python2-requests-oauthlib` can be dropped because it's pulled in automatically.

rabajaj0509 replied:
@ekohl good catch! I checked for Fedora, it comes as an installing dependency for the `keycloak-httpd-client-install` package.

ekohl replied:
Could you have a look at this?
```suggestion
# yum install mod_auth_openidc keycloak-httpd-client-install -y
```

apurvabhide replied:
@ekohl , I have updated this command.

ekohl commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,106 @@
+# export KEYCLOAK_URL=https://keycloak.example.com
+# export KEYCLOAK_REALM=ssl-realm
+# export KEYCLOAK_USER=admin
+# keycloak-httpd-client-install --app-name foreman-openidc --keycloak-server-url $KEYCLOAK_URL --keycloak-admin-username $KEYCLOAK_USER --keycloak-realm $KEYCLOAK_REALM --keycloak-admin-realm master --keycloak-auth-role root-admin -t openidc -l /users/extlogin -d
```
This appears to fail for me. First I had to install python-lxml:
```
[root@68e8ec694948 /]# keycloak-httpd-client-install --help
Traceback (most recent call last):
  File "/usr/bin/keycloak-httpd-client-install", line 8, in <module>
    from lxml import etree
ImportError: No module named lxml
```

After that it doesn't recognize `-t`

```
[root@68e8ec694948 /]# export KEYCLOAK_URL=https://keycloak.example.com
[root@68e8ec694948 /]# export KEYCLOAK_REALM=ssl-realm
[root@68e8ec694948 /]# export KEYCLOAK_USER=admin
[root@68e8ec694948 /]# keycloak-httpd-client-install --app-name foreman-openidc --keycloak-server-url $KEYCLOAK_URL --keycloak-admin-username $KEYCLOAK_USER --keycloak-realm $KEYCLOAK_REALM --keycloak-admin-realm master --keycloak-auth-role root-admin -t openidc -l /users/extlogin -d
usage: keycloak-httpd-client-install [-h] [--no-root-check] [-v] [-d]
                                     [--show-traceback] [--log-file LOG_FILE]
                                     --app-name APP_NAME [--force]
                                     [--permit-insecure-transport]
                                     [--tls-verify TLS_VERIFY]
                                     [--template-dir TEMPLATE_DIR]
                                     [--httpd-dir HTTPD_DIR] -r KEYCLOAK_REALM
                                     -s KEYCLOAK_SERVER_URL
                                     [-a {root-admin,realm-admin,anonymous}]
                                     [-u KEYCLOAK_ADMIN_USERNAME]
                                     [-P KEYCLOAK_ADMIN_PASSWORD_FILE]
                                     [-p DEPRECATED_KEYCLOAK_ADMIN_PASSWORD]
                                     [--keycloak-admin-realm KEYCLOAK_ADMIN_REALM]
                                     [--initial-access-token INITIAL_ACCESS_TOKEN]
                                     [--client-originate-method {descriptor,registration}]
                                     [--mellon-key-file MELLON_KEY_FILE]
                                     [--mellon-cert-file MELLON_CERT_FILE]
                                     [--mellon-hostname MELLON_HOSTNAME]
                                     [--mellon-https-port MELLON_HTTPS_PORT]
                                     [--mellon-root MELLON_ROOT]
                                     [--mellon-endpoint MELLON_ENDPOINT]
                                     [--mellon-entity-id MELLON_ENTITY_ID]
                                     [--mellon-idp-attr-name MELLON_IDP_ATTR_NAME]
                                     [--mellon-organization-name MELLON_ORGANIZATION_NAME]
                                     [--mellon-organization-display-name MELLON_ORGANIZATION_DISPLAY_NAME]
                                     [--mellon-organization-url MELLON_ORGANIZATION_URL]
                                     [-l MELLON_PROTECTED_LOCATIONS]
keycloak-httpd-client-install: error: unrecognized arguments: -t openidc
```

rabajaj0509 replied:
@ekohl  which version of `mod_auth_openidc` and `keycloak-httpd-client-install`
 are you using?

ekohl replied:
Looks like currently CentOS ships the following:
* keycloak-httpd-client-install.noarch 0:0.8-1.el7
* mod_auth_openidc.x86_64 0:1.8.8-5.el7

gnurag replied:
keycloak-httpd-client-install.noarch version `0:0.8-1.el7` could be the reason for failure. I reckon OpenID Connect support was added to keycloak-httpd-client-install in version 1.x. 

@rahulbajaj0509 Did you test this setup with a custom built keycloak-httpd-client-install?
@upadhyeammit FYI

rabajaj0509 replied:
@gnurag that is correct, keycloak-httpd-client-install package with version >= 1.x. supports OpenID Connect.

@ekohl can we have this bit merged for now, so that we have some documentation about the feature rather than nothing? and then we can make improvements to it as when we face issues.
I am completely ok to not merge this if you believe so.

ekohl replied:
I would add a note that you need at least version 1.x and the one in base is too old.

rabajaj0509 replied:
Makes sense.

apurvabhide replied:
I have added a note for keycloak-httpd-client-install package version.

ekohl commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
````diff
@@ -0,0 +1,106 @@
+
+2. Log in to Foreman:
+```
+hammer auth login oauth --oidc-token-endpoint https://keycloak.example.com/token --oidc-client-id foreman-client --username rabajaj --password changeme
````
I'd avoid personal names in examples. Safer to have something as `myuser`

gnurag replied:
Please read comment on `--oidc-client-id foreman-client` above.

ekohl commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
````diff
@@ -0,0 +1,106 @@
+
+2. Log in to Foreman:
+```
+hammer auth login oauth --two-factor --oidc-token-endpoint https://keycloak.exmaple.com/token --oidc-authorization-endpoint https://keycloak.example.com/auth --oidc-client-id foreman-client --oidc-redirect-uri urn:ietf:wg:oauth:2.0:oob
````
```suggestion
hammer auth login oauth --two-factor --oidc-token-endpoint https://keycloak.example.com/token --oidc-authorization-endpoint https://keycloak.example.com/auth --oidc-client-id foreman-client --oidc-redirect-uri urn:ietf:wg:oauth:2.0:oob
```

gnurag replied:
@rahulbajaj0509 Just a note for hammer implementation: `urn:ietf:wg:oauth:2.0:oob` is a standard redirect string and the user need not remember it. This redirect string is always used when `--two-factor` is passed, so can this be made implicit?

gnurag replied:
Does `--oidc-client-id foreman-client` in this line correspond to `keycloak-httpd-client-install --app-name foreman-openidc` from line 35? If they're same, perhaps use the same names to avoid confusion. If not, please ignore.

gnurag commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
````diff
@@ -0,0 +1,106 @@
+
+2. Register Katello/Foreman as application in the Keycloak:
+```
+# export KEYCLOAK_URL=https://keycloak.example.com
````
On line 20, Keycloak URL is set to https://sso.example.com. Please update this (or the previous one?) URL to maintain consistency.

gnurag commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
````diff
@@ -0,0 +1,106 @@
+# export KEYCLOAK_USER=admin
+# keycloak-httpd-client-install --app-name foreman-openidc --keycloak-server-url $KEYCLOAK_URL --keycloak-admin-username $KEYCLOAK_USER --keycloak-realm $KEYCLOAK_REALM --keycloak-admin-realm master --keycloak-auth-role root-admin -t openidc -l /users/extlogin -d
+```
+Warning: This feature is not yet supported by foreman-installer. As a result, re-running the foreman-installer command can delete the changes in Apache files added by this feature.
````
can delete the changes in Apache files added by this feature -> can purge the changes in Apache files added by keycloak-httpd-client-install

gnurag commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,106 @@
+
+5. Set the following OIDC parameters:
+
+ **OIDC Algorithm**: Algorithm type with which JWT was encoded in the OpenID provider. For example, For example, set this parameter to RS-256.
```
For example, For example -> For example,

gnurag commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,106 @@
+
+ **OIDC Algorithm**: Algorithm type with which JWT was encoded in the OpenID provider. For example, For example, set this parameter to RS-256.
+
+ **OIDC Audience**: Name of the OIDC audience used for authentication. In case of [Keycloak](https://www.keycloak.org/), it is Client ID.
```
Do we need to link to Keycloak project's website again? We've already provided its link in the introduction at the top. What convention do other manuals follow?

gnurag commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,106 @@
+
+An end user can perform SSO by the following ways:
+* User Interface
+* Hammer-CLI
```
Perhaps add a section to demonstrate SSO workflow for Foreman API.

apurvabhide replied:
@gnurag , as discussed with @rahulbajaj0509 , I will add it in another PR in 5.1 API section .

gnurag commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,106 @@
+
+* Log in to the https://foreman.example.com/users/extlogin
+
+ If credentials are correct, it redirects to foreman dashboard.
```
++ Foreman will automatically redirect to Keycloak Sign-in page. If credentials are correct, it redirects to Foreman dashboard.

gnurag commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,106 @@
+Authorization Code Flow is a two step process:
+
+* Authorization Code Flow returns an Authorization Code to the client.
+* this Code can be exchanged for an access token directly.
```
s/this/This/

gnurag reviewed (commented):
In general, please also pay attention to using appropriate capitalization.

apurvabhide commented:
> In general, please also pay attention to using appropriate capitalization.

@gnurag, Done, I have updated both the files with your suggestions.

ekohl commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,107 @@
+### Single Sign-on for Foreman using OpenID Connect protocol
+
+An OpenID Provider(OP) implements a Single Sign-on (SSO) using an OpenID Connect (OIDC) protocol. OIDC verifies the identity of the End-user depending on the given authentication details. In this example, [Keycloak](https://www.keycloak.org/) is used as the OpenID provider, and Foreman is the Relying Party (RP)
```
```suggestion
An OpenID Provider(OP) implements a Single Sign-on (SSO) using an OpenID Connect (OIDC) protocol. OIDC verifies the identity of the End-user depending on the given authentication details. In this example, [Keycloak](https://www.keycloak.org/) is used as the OpenID provider, and Foreman is the Relying Party (RP).
```

ekohl commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,107 @@
+Prerequisites
+
+* A working installation of Foreman at https://foreman.example.com.
+* A working installation of OpenID provider, for example, Keycloak at https://keycloak.example.com, which uses OIDC protocol. Ensure that the Keycloak server is running.
```
```suggestion
* A working installation of OpenID provider, for example, Keycloak at https://keycloak.example.com, which uses the OIDC protocol. Ensure that the Keycloak server is running.
```

ekohl commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,107 @@
+
+* Log in to the https://foreman.example.com/users/extlogin
+
+ Foreman will automatically redirect to Keycloak Sign-in page. If credentials are correct, it redirects to Foreman dashboard.
```
```suggestion
Foreman will automatically redirect to Keycloak Sign-in page. If credentials are correct, it redirects to the Foreman dashboard.
```

apurvabhide replied:
For the indentation purpose, I have added one space before this sentence.

ekohl commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
````diff
@@ -0,0 +1,107 @@
+Procedure
+
+1. Install the mod_auth packages:
+```
````
We can use console syntax highlighting (and lower is another block which can use this)
```suggestion
```console
```

apurvabhide replied:
done, I have added ```console for the console syntax highlighting.

ekohl reviewed (approved):
Small nitpicks that would be nice to fix but overall :+1: to merging.

rabajaj0509 commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,107 @@
+
+5. Set the following OIDC parameters:
+
+ **OIDC Algorithm**: Algorithm type with which JWT was encoded in the OpenID provider. For example, set this parameter to RS-256.
```
I think the parameter is `RS256`

apurvabhide replied:
done, I have updated this.

rabajaj0509 commented:
Thanks @apurvabhide for all the hard work! merging :)

apurvabhide commented:
> Thanks @apurvabhide for all the hard work! merging :)

My pleasure :) 
Thank you all, for reviewing it.

ekohl commented on _includes/manuals/1.24/5.7.7_openid_sso.md:
```diff
@@ -0,0 +1,107 @@
+Warning: This feature is not yet supported by foreman-installer. As a result, re-running the foreman-installer command can purge the changes in Apache files added by the keycloak-httpd-client-install.
+
+3. Log in to the Foreman admin page and click **Administer** → **Settings** → **Authentication**.
+ ![Foreman UI](Foreman_UI.png)
```
I had a look at this and this image is not included leading to a HTTP 404. This is true for other images as well.

apurvabhide replied:
@ekohl , apologies for the late reply. Could you please help me with the image folder, should I add images in the /static/images/ directory?
Thanks
